tests/framework/db/ActiveRecordTest.php

```diff
@@ -2192,4 +2192,78 @@ public function testVirtualRelation()
         $this->assertNotNull($order->virtualCustomer);
     }
 
+    public function testGetAttributeLabel()
+    {
+        // Model 2 and 3 are represented by objects.
+        $model1 = new LabelTestModel1();
+        $model2 = new LabelTestModel2();
+        $model3 = new LabelTestModel3();
+        $model2->populateRelation('model3', $model3);
+        $model1->populateRelation('model2', $model2);
+        $this->runTestGetAttributeLabelForModel($model1);
+
+        // Model 2 and 3 are represented by arrays instead of objects.
+        $model2 = ['model3' => []];
+        $model1->populateRelation('model2', $model2);
+        $this->runTestGetAttributeLabelForModel($model1);
+    }
+
+    private function runTestGetAttributeLabelForModel($model)
```

Comment:
Running the same test with different data is usually done with [data providers](https://phpunit.de/manual/3.7/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers)

Reply:
Can't say this feature of PHPUnit increases test code readability (rather the opposite) but since it seems to be used everywhere in Yii2 tests I changed the test code to use data providers too (for consistency reasons).